File: backend/src/main/java/com/ecms/service/PayrollService.java
package com.ecms.service;

import com.ecms.dto.request.PayrollItemUpdateRequest;

import java.util.List;
import java.util.Map;

/**
 * @author  ThangNB - HE201024
 * @created 2026-07-19
 * @updated 2026-07-19
 *
 * Payroll approval contract, one period per calendar month
 * (UC-54 Approve Payroll).
 *
 * Business rules: BR-17 (only the Clinic Manager may approve a payroll
 * period), BR-09 (once APPROVED, line items are locked and never deleted).
 */
public interface PayrollService {

    /**
     * Generates — or regenerates — the DRAFT payroll for a period from the
     * activity data already captured by the system (UC-54 normal flow step 2).
     *
     * @param year  pay period year
     * @param month pay period month, 1-12
     * @return the draft period with its line items
     *
     * Validate: BR-09 / UC-54 POST-2 — an already APPROVED period must not be
     * regenerated, since its lines are locked.
     */
    Map<String, Object> generateDraft(int year, int month);

    /**
     * Lists pay periods, newest first.
     *
     * @return period summaries
     */
    List<Map<String, Object>> listPeriods();

    /**
     * Loads one pay period together with its line items.
     *
     * @param periodId pay period primary key
     * @return the period and its lines
     */
    Map<String, Object> getPeriod(Long periodId);

    /**
     * Adjusts a single payroll line, with a justification note
     * (UC-54 normal flow step 3).
     *
     * @param itemId  payroll line primary key
     * @param request the amounts to override plus the reason
     * @return the updated line
     *
     * Validate: BR-09 — editing is only permitted while the period is DRAFT;
     * once APPROVED the lines are locked (UC-54 POST-2).
     */
    Map<String, Object> updateItem(Long itemId, PayrollItemUpdateRequest request);

    /**
     * Approves the payroll period: status → APPROVED, all lines locked, event
     * written to the Audit Log (UC-54 normal flow steps 4-7).
     *
     * @param periodId    pay period primary key
     * @param actorUserId the approving manager, recorded in the audit entry
     * @return the approved period
     *
     * Validate: BR-17 — approval authority is restricted to the Clinic
     * Manager; BR-09 — approval is irreversible, lines can no longer be edited
     * or removed.
     */
    Map<String, Object> approve(Long periodId, Long actorUserId);
}

File: frontend/src/services/payrollService.js
/**
 * @author  ThangNB - HE201024
 * @created 2026-07-19
 * @updated 2026-07-19
 *
 * API client for payroll approval (UC-54 Approve Payroll).
 *
 * Validate: BR-17 (only the Clinic Manager may approve) and BR-09 (approved
 * lines are locked) are both enforced server-side; failures surface here as
 * error responses rather than being pre-checked in the UI.
 */
import axiosClient from '../api/axiosClient'

export const payrollService = {
  /**
   * Generates or regenerates the DRAFT payroll for a month.
   * @param {number} year  pay period year
   * @param {number} month pay period month, 1-12
   * @returns {Promise} the draft period with its lines
   */
  generate: (year, month) => axiosClient.post('/v1/payroll/generate', null, { params: { year, month } }),

  /**
   * Lists pay periods, newest first.
   * @returns {Promise} period summaries
   */
  listPeriods: () => axiosClient.get('/v1/payroll/periods'),

  /**
   * Loads one period with all its payroll lines.
   * @param {number} id pay period id
   * @returns {Promise} the period and its lines
   */
  getPeriod: (id) => axiosClient.get(`/v1/payroll/periods/${id}`),

  /**
   * Applies a Manager override to one payroll line.
   * @param {number} id payroll line id
   * @param {{baseSalary?:number, performanceBonus?:number, deduction?:number, note?:string}} data
   * @returns {Promise} the updated line
   *
   * Validate: BR-09 — rejected by the backend once the period is APPROVED.
   */
  updateItem: (id, data) => axiosClient.patch(`/v1/payroll/items/${id}`, data),

  /**
   * Approves a pay period, locking every line (UC-54 step 4).
   * @param {number} id pay period id
   * @returns {Promise} the approved period
   *
   * Validate: BR-17 — the approving manager is taken from the auth token
   * server-side, not sent from here.
   */
  approve: (id) => axiosClient.post(`/v1/payroll/periods/${id}/approve`),
}
